feat: add chunk storage metrics

tustvold opened the pull request:
This adds two new gauges that are aggregated for each table

* catalog_chunks_storage - the number of chunks in the given storage
* catalog_row_count - the number of rows in the given storage

These should be helpful for understanding how data flows through the system. We currently have `catalog_chunks_total` but this is a counter and somewhat hard to interpret.

e-dard commented on server/src/db.rs:
```diff
@@ -1456,7 +1456,22 @@ mod tests {
+                .has_metric_family(name)
+                .with_labels(&[
+                    ("db_name", "placeholder"),
+                    ("source", source),
```
I wonder if `location` might be a better tag name?

e-dard commented on server/src/db/catalog/metrics.rs:
```diff
@@ -56,8 +62,24 @@ impl CatalogMetrics {
         );
 
+        let storage_gauge = self.metrics_domain.register_gauge_metric_with_labels(
+            "chunks_storage",
```
What about something like `loaded` for the metric prefix and then `Some("chunks")` for then `unit`? Then the full metric could be something like:

```
catalog_loaded_chunks {db_name="foo", locaction="read_buffer", svr_id="42", table="m1"} 22
catalog_loaded_chunks {db_name="foo", locaction="mutable_buffer", svr_id="42", table="m1"} 2
```

And so on. Seems to flow a bit better to me.

e-dard commented on server/src/db/catalog/metrics.rs:
```diff
@@ -56,8 +62,24 @@ impl CatalogMetrics {
+        let storage_gauge = self.metrics_domain.register_gauge_metric_with_labels(
+            "chunks_storage",
+            None,
+            "The number of chunks in a given storage",
```
```suggestion
            "The number of chunks loaded in a each chunk storage location",
```

crepererum commented on server/src/db/catalog/metrics.rs:
```diff
@@ -56,8 +62,24 @@ impl CatalogMetrics {
         );
 
+        let storage_gauge = self.metrics_domain.register_gauge_metric_with_labels(
+            "chunks_storage",
```
```suggestion
            "chunks_storage_count",
```

IIRC a metric should always reflect it's unit to some degree.

e-dard replied:
I addressed this in my review.

e-dard commented on server/src/db/catalog/metrics.rs:
```diff
@@ -56,8 +62,24 @@ impl CatalogMetrics {
+        );
+
+        let row_gauge = self.metrics_domain.register_gauge_metric_with_labels(
+            "row_count",
```
I wonder if this could be `loaded` for the prefix name and then `Some("rows")` for the unit? The full metric would then be:

```
catalog_loaded_rows{db_name="foo",location="mutable_buffer",svr_id="42",table="m0"} 99999
```

It would also sit nicely next to `catalog_loaded_chunks` in the exposed metrics endpoint.

e-dard commented on server/src/db/catalog/metrics.rs:
```diff
@@ -56,8 +62,24 @@ impl CatalogMetrics {
+        let row_gauge = self.metrics_domain.register_gauge_metric_with_labels(
+            "row_count",
+            None,
+            "The number of row in a given storage",
```
```suggestion
            "The number of rows loaded in each chunk storage location",
```

crepererum commented on server/src/db/catalog/metrics.rs:
```diff
@@ -139,64 +175,86 @@ impl PartitionMetrics {
     }
 }
 
+/// Created from a `metrics::Gauge` and extracts a `GaugeValue` for each chunk storage
```
Maybe mention that this object should be used once per chunk and internally reflects the three storage types. I got a bit confuser how this was meant to be used (looking at the code within the catalog chunk clarified that).

crepererum reviewed (approved):
Small nitpicks, otherwise OK

e-dard commented on server/src/db.rs:
```diff
@@ -1473,11 +1488,18 @@ mod tests {
 
+        assert_metric("catalog_chunks_storage", "mutable_buffer", 1.0);
+        assert_metric("catalog_chunks_storage", "read_buffer", 0.0);
+        assert_metric("catalog_chunks_storage", "parquet", 0.0);
```
Instead of `parquet` could we use `object store` or similar? I'm just thinking that if/when we add a local parquet file cache it will be a different parquet-backed chunk location and it would be good to be able to add that without renaming the existing metric labels.

e-dard reviewed (commented):
I have tested this locally and it seems to work perfectly. I do have some thoughts about the names of the metrics though.

alamb commented:
Thank you for this PR @tustvold

tustvold commented:
I've renamed the metrics as requested.

NB: I've also renamed source -> location and parquet -> object_store, for the memory metrics as well for consistency.

PTAL

e-dard reviewed (approved):
😍
